rep-2012.rst

```diff
@@ -0,0 +1,466 @@
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 07-Jun-2022
+Post-History:
+
+Abstract
+========
+
+This REP proposes a feature to introspect ROS services during runtime.
+The feature allows users to remotely monitor service requests and responses.
+
+
+Terminology
+===========
+
+:Service:
+  A ROS service.
+  Accepts requests from clients and sends responses.
+:Client:
+  A ROS service client.
+  Sends requests to services and receives responses.
+:Request:
+  A ROS service request.
+:Response:
+  A ROS service response.
+
+
+Motivation
+==========
+
+The primary motivation for this proposal is to make it easier for users to externally validate that services are operating as expected.
+Drawing an analogy to ROS topics, there exists tools and libraries for "echoing" and recording messages sent over a topic, and we would like to see the same kind of capabilities for services.
+Specifically, we want the capability to introspect requests and responses that are sent to and from services and clients.
+Being able to remotely monitor services allows users to more effectively troubleshoot issues in a ROS system.
+For example, we could verify requests are being received by a service by employing a command-line tool during runtime.
+Or, we could post-process recorded requests and responses to validate their content.
+
+There are additional features that could leverage this proposal, such as:
+
+- Playback of recorded services (for example, from a rosbag [1]_)
+- Introspection of ROS actions, which are built on services
+- Validation of a live ROS system by referencing a recording from a previous session
+
+Though this proposal focuses on the core feature of introspecting requests and responses, we want to keep the design flexible so additional features like those listed above can be implemented in the future.
+
+
+Specification
+=============
+
+Publishing Service Events
+-------------------------
+
+Whenever a request or response is sent or received, a *service event* message will be published to a topic.
+Services are responsible for publishing a message when they receive a request and when they send a response.
+Likewise, clients are responsible for publishing a message when they send a request and when they receive a response.
+Therefore, we have a total of four possible events:
+
+:Request Sent:
+  Emitted from a client after sending a request to a service.
+:Request Received:
+  Emitted from a service after receiving a request from a client.
+:Response Sent:
+  Emitted from a service after sending a response to a client.
+:Response Received:
+  Emitted from a client after receiving a response from a service.
+
+Request event messages shall be published to the hidden topic ``/SERVICE_NAME/_request_event``, where ``SERVICE_NAME`` is the fully-qualified name of the service.
+Similarly, response event messages shall be published to the hidden topic ``/SERVICE_NAME/_response_event``.
```

Comment:
If we use a single topic (as is proposed below) - would we then get 4 messages to that topic per service call? Request Sent, Request Received, Response Sent, Response Received?

Reply:
Yes, I would think so. Users would have to check the type in the message to determine the source of the event.